task-impls/src/consensus.rs

```diff
@@ -263,51 +266,80 @@ where
         };
         Some(leaf.clone())
     }
-
+    async fn vote_if_able(&self) {
+        if let Some(proposal) = &self.current_proposal {
+            if let Some(cert) = self.certs.get(&proposal.get_view_number()) {
+                let view = cert.view_number;
+                let vote_token = self.quorum_exchange.make_vote_token(view);
+                // TODO: do some of this logic without the vote token check, only do that when voting.
+                match vote_token {
+                    Err(e) => {
+                        error!("Failed to generate vote token for {:?} {:?}", view, e);
+                    }
+                    Ok(None) => {
+                        info!("We were not chosen for consensus committee on {:?}", view);
+                    }
+                    Ok(Some(vote_token)) => {
+                        // Validate the DAC.
+                        if !self
+                            .committee_exchange
+                            .is_valid_cert(&cert, proposal.block_commitment)
+                        {
+                            warn!("Invalid DAC in proposal! Skipping proposal.");
+                            let _message: GeneralConsensusMessage<TYPES, I> =
+                                self.quorum_exchange.create_no_message(
+                                    proposal.justify_qc.commit(),
+                                    proposal.justify_qc.leaf_commitment,
+                                    cert.view_number,
+                                    vote_token,
+                                );
+                        } else {
+                            let _message: GeneralConsensusMessage<TYPES, I> =
+                                self.quorum_exchange.create_yes_message(
+                                    proposal.justify_qc.commit(),
+                                    proposal.justify_qc.leaf_commitment,
+                                    cert.view_number,
+                                    vote_token,
+                                );
```

Comment:
What are these messages used for? Do we need to hook this into something?

Reply:
yep, we should emit this as an event, I think I missed this when I moved the logic around, good catch.